frame/election-provider-multi-phase/test-staking-e2e/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+        )
+    };
+}
+
+fn log_current_time() {
+	log!(
+		trace,
+		"block: {:?}, session: {:?}, era: {:?}, EPM phase: {:?} ts: {:?}",
+		System::block_number(),
+		Session::current_index(),
+		Staking::current_era(),
+		ElectionProviderMultiPhase::current_phase(),
+		Timestamp::now()
+	);
+}
+
+#[test]
+fn setup_works() {
+	ExtBuilder::default().initialize_first_session(true).build_and_execute(|| {
+		assert_eq!(active_era(), 0);
+
+		assert!(start_next_active_era().is_ok());
+		assert!(start_next_active_era().is_ok());
+		assert_eq!(active_era(), 2);
+
+		// if the solution is delayed, EPM will end up in emergency mode and eras won't progress.
+		assert!(start_next_active_era_delayed_solution().is_ok());
+		assert!(start_next_active_era_delayed_solution().is_ok());
+		assert!(ElectionProviderMultiPhase::current_phase().is_emergency());
+		assert_eq!(active_era(), 3);
+	});
+}
+
+#[test]
+/// Replicates the Kusama incident of 8th Dec 2022 and its resolution through the governance
+/// fallback.
+///
+/// After enough slashes to exceed the `Staking::OffendingValidatorsThreshold`, the staking pallet
+/// set `Forcing::ForceNew`. When a new session starts, staking will start to force a new era and
+/// calls <EPM as election_provider>::elect(). If at this point EPM and the staking miners did not
+/// have enough time to queue a new solution (snapshot + solution submission), the election request
+/// fails. If there is no election fallback mechanism in place, EPM enters in emergency mode.
+/// Recovery: Once EPM is in emergency mode, subsequent calls to `elect()` will fail until a new
+/// solution is added to EPM's `QueuedSolution` queue. This can be achieved through
+/// `Call::set_emergency_election_result` or `Call::governance_fallback` dispatchables. Once a new
+/// solution is added to the queue, EPM phase transitions to `Phase::Off` and the election flow
+/// restarts.
+fn enters_emergency_phase_after_forcing_before_elect() {
+	ExtBuilder::default().initialize_first_session(true).build_and_execute(|| {
+		log!(
+			trace,
+			"current validators (staking): {:?}",
+			<Runtime as pallet_staking::SessionInterface<AccountId>>::validators()
+		);
+		let session_validators_before = Session::validators();
+
+		roll_to_epm_off();
+		assert!(ElectionProviderMultiPhase::current_phase().is_off());
+		log_current_time();
+
+		assert_eq!(pallet_staking::ForceEra::<Runtime>::get(), pallet_staking::Forcing::NotForcing);
+		// slashes so that staking goes into `Forcing::ForceNew`.
+		slash_through_offending_threshold();
+
+		assert_eq!(pallet_staking::ForceEra::<Runtime>::get(), pallet_staking::Forcing::ForceNew);
+
+		advance_session_delayed_solution();
+		assert!(ElectionProviderMultiPhase::current_phase().is_emergency());
+		log_current_time();
+
+		// try to advance 2 eras with a delayed solution.
+		assert!(start_next_active_era_delayed_solution().is_ok());
+		assert!(start_next_active_era_delayed_solution().is_ok());
```

Comment:
May be we should assert era is not changing?

Reply:
Done!